### PR TITLE
Dev/kpf

### PR DESCRIPTION
**Pull request by collinsr**
This branch handles KPF I/O for canonical data types useful for object and activity detection and recognition. It has some examples which I'm hoping to convert to test cases.

This branch requires YAML, which is currently landing in fletch on a different pull request.

### CODE REVIEW

**Comment by aaron-bray** [comment]
It looks like this is the bare bones i/o classes
It has support for breaking out basic types, like a bounding box and polygons

Are there plans to create a yaml file from an actual detected_object_set object?
Are we still going to need to create that link between vital types generated from a pipeline to this yaml format?

**Comment by collinsr** [comment]
You are correct; this reads and writes the basic objects (for example, the V1 YAML annotations) into the canonical_io_types. Next step would be to have another interface that maps into vital types (much like the examples/kpf_example_*.cxx files do.) I'm hoping that will be fairly straightforward.

**Comment by aaron-bray** [comment]
Yeah, the examples help
We don't need to worry about reading yaml back into a vital type, just vital type out to yaml (at this point) (right?)

The examples are handy for non kwiver users to see and use.

It might also be worth having actual valid yaml files that contain all data needed for an evaluation somewhere, just in case there is a performer that whats to write their own writer, they have an example to copy off of. (as silly as that would be)

**Comment by collinsr** [comment]
KWant will have to read it for scoring.

Once I get these draft slides out the door I'm going to create test cases out of the examples and check some of the DIVA annotations in with those.

File: vital/exceptions/kpf.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+ */
+class VITAL_EXPORT kpf_token_underrun_exception
+  : public kpf_exception
+{
+public:
+  /// Constructor
+  /**
+   * \param message     Description of the parsing circumstances
+   */
+  kpf_token_underrun_exception(std::string message) VITAL_NOTHROW;
```
[the remark this answers is not in the document]
Comment (aaron-bray) [reply vital/exceptions/kpf.h]: Yeah, const std::string& should be default for string passing.

I will double check these, I am getting to issues fixed on windows I am running into

I will probably end up creating a new PR from my branch when I get everything building

[the remark this answers is not in the document]
Comment (mwoehlke-kitware) [reply vital/exceptions/kpf.h]: nit: `std::string const&` 😉.

**Comment by mwoehlke-kitware** [comment]
> It has some examples which I'm hoping to convert to test cases.

Indeed, there seems to be more than enough code here that some (Google Test) unit tests would be nice...

File: vital/exceptions/kpf.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (aaron-bray) [reply vital/exceptions/kpf.h]: hard g or soft g in gif?

[the remark this answers is not in the document]
Comment (mwoehlke-kitware) [reply vital/exceptions/kpf.h]: ...depends on my mood :wink:.

But seriously, note that I ended that comment with a wink. (That said, it *would* be nice if we were consistent, and I'm sure I don't need to mention my preference. As an *explanation*, however, I offer `const LPSTR`, by which a certain library — not one of ours, thankfully — which I will not name had horribly broken C bindings. Also, doing a find-and-replace `T*` → `T_ptr` is dangerously broken with `const T*`... which may be what caused said brokenness.)

**Comment by aaron-bray** [comment]
Yeah, I saw

I am impartial, just want to be consistent. I like your explanation. I do
enjoy programming nit picks 😁

**Comment by mwoehlke-kitware** [comment]
> I am impartial, just want to be consistent.

Indeed; we are not currently consistent. (My *very* quick-n-dirty count, with regexes that may have false positives and only consider pointers/references, actually show a slight bias — about 3:2 — toward `T const`.)

**Comment by dstoup** [comment]
Jenkins test this please

**Comment by aaron-bray** [comment]
http://proteus:8080/job/KwiverLinuxPR/75/ and message: 'Build started
sha1 is merged.'
FileNotFoundException means that the credentials Jenkins is using is
probably wrong. Or the user account does not have write access to the
repo.

**Comment by aaron-bray** [comment]
This is the Jenkins kwiver only building arrows with out anything else.

It does not know where boost is in that case

**Pull request merged**